resilience4j-kotlin/src/main/kotlin/io/github/resilience4j/kotlin/circuitbreaker/FlowCircuitBreaker.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ *
+ *  Copyright 2019 authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.kotlin.circuitbreaker
+
+import io.github.resilience4j.circuitbreaker.CallNotPermittedException
+import io.github.resilience4j.circuitbreaker.CircuitBreaker
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emitAll
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.onCompletion
+import java.util.concurrent.CancellationException
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.coroutineContext
+
+/**
+ * Wraps the given flows collection with a permission check on the supplied [circuitBreaker].
+ *
+ * The events are published to the circuit breaker via the `onCompletion` operator chained to the given flow.
+ *
+ * When the circuit breaker is OPEN, flow collection throws a [CallNotPermittedException]
+ * ```
+ * flowOf("a", "b", "c")
+ *     .circuitBreaker(circuitBreaker)
+ *     .collect { println(it) } // Throws CallNotPermittedException
+ * ```
+ *
+ * Coroutine cancellation (_normal_ and _exceptional_) doe not record any events on the circuit breaker
+ * and the acquired permission is released.
+ *
+ */
+@UseExperimental(ExperimentalCoroutinesApi::class)
+fun <T> Flow<T>.circuitBreaker(circuitBreaker: CircuitBreaker): Flow<T> =
+    flow {
```

Comment:
I think you can use [`onStart`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/on-start.html) here.
```
onStart {
    if(circuitBreaker.tryAcquirePermission()) {
        val start = System.nanoTime()
        val source = this@circuitBreaker.onCompletion { e ->
            when {
                isCancellation(e, coroutineContext) -> circuitBreaker.releasePermission()
                e == null -> circuitBreaker.onSuccess(System.nanoTime() - start, TimeUnit.NANOSECONDS)
                else -> circuitBreaker.onError(System.nanoTime() - start, TimeUnit.NANOSECONDS, e)
            }
        }
    } else {
        throw CallNotPermittedException.createCallNotPermittedException(circuitBreaker)
    }
}
```

Reply:
The completion handler won’t be subscribed to since its being appended in the onstart. Val source goes used here.

Reply:
This is what I meant.
```
fun <T> Flow<T>.circuitBreaker(circuitBreaker: CircuitBreaker): Flow<T> = onStart {
    if(circuitBreaker.tryAcquirePermission()) {
        val start = System.nanoTime()
        val source = this@circuitBreaker.onCompletion { e ->
            when {
                isCancellation(e, coroutineContext) -> circuitBreaker.releasePermission()
                e == null -> circuitBreaker.onSuccess(System.nanoTime() - start, TimeUnit.NANOSECONDS)
                else -> circuitBreaker.onError(System.nanoTime() - start, TimeUnit.NANOSECONDS, e)
            }
        }
    } else {
        throw CallNotPermittedException.createCallNotPermittedException(circuitBreaker)
    }
}
```

The [implementation](https://github.com/Kotlin/kotlinx.coroutines/blob/7f0da424ed98e3a30e0a7ca2daff33f9a9cdbf0c/kotlinx-coroutines-core/common/src/flow/operators/Emitters.kt#L71) is the same, except with some optimisations.

Reply:
Ah I see. Using `onCompletion` in there is a bit awkward to read I think. Perhaps just using try /catch /finally would be clearer?

Reply:
The 'onCompletion' operator will return a new instance of flow not the original receiver. Running the tests against this implementation fails. To use 'onStart' we would need to structure the method a little differently so that the completion operator is still chained.

```kotlin
fun <T> Flow<T>.circuitBreaker(circuitBreaker: CircuitBreaker): Flow<T> {
    var start: Long = 0
    return onStart {
        start = System.nanoTime()
        if (!circuitBreaker.tryAcquirePermission()) {
            throw CallNotPermittedException.createCallNotPermittedException(circuitBreaker)
        }
    }.onCompletion { e ->
        when {
            isCancellation(e, coroutineContext) -> circuitBreaker.releasePermission()
            e == null -> circuitBreaker.onSuccess(System.nanoTime() - start, TimeUnit.NANOSECONDS)
            else -> circuitBreaker.onError(System.nanoTime() - start, TimeUnit.NANOSECONDS, e)
        }
    }
}
``` 

But I think this implementation would go against the patterns defined in the stdlib for implementing [opertators](https://github.com/Kotlin/kotlinx.coroutines/tree/master/kotlinx-coroutines-core/common/src/flow/operators). 

Most operators that transform their receiver do so by encapsulating their implementation in a flow builder ie.`flow{}`. With the example above we could potentially see issues with the setting of the start val and reading in the completion handler

Reply:
```
fun <T> Flow<T>.circuitBreaker(circuitBreaker: CircuitBreaker): Flow<T> = flow {
    if (circuitBreaker.tryAcquirePermission()) {
        val start = System.nanoTime()
        try {
            emitAll(this@circuitBreaker)
            
            circuitBreaker.onSuccess(System.nanoTime() - start, TimeUnit.NANOSECONDS)
        } catch (e: Throwable) {
            if (isCancellation(e, coroutineContext)) {
                circuitBreaker.releasePermission()
                throw e
            } else {
                circuitBreaker.onError(System.nanoTime() - start, TimeUnit.NANOSECONDS, e)
            }
        }
    } else {
        throw CallNotPermittedException.createCallNotPermittedException(circuitBreaker)
    }
}
```

Reply:
After reading the doc for `releasePermission`, I'm not sure it should be called based on cancellation, since you don't really know if cancellation happened before use or not.

Reply:
If the flow is already completed then the on complete block wont be invoked again. Meaning that if a cancellation were to occur after the circuit breaker was used, it would not affect completed flows.
 
As for using a try / catch block, the flow documentation explicitly advises against this. Since it violates [Exception Transparency](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/#exception-transparency)
>Flow implementations never catch or handle exceptions that occur in downstream flows. From the implementation standpoint it means that calls to emit and emitAll shall never be wrapped into try { ... } catch { ... } blocks.

Reply:
I guess there's no nice way to write this then.

Reply:
> After reading the doc for releasePermission, I'm not sure it should be called based on cancellation, since you don't really know if cancellation happened before use or not.

Your implementation is fine. The method `circuitBreaker.releasePermission()` should be invoked, when the flow is cancelled, but neither `circuitBreaker.onSuccess` nor `circuitBreaker.onError` can be called.

Reply:
The current implementation will only ever call one of `releasePermission`, `onSuccess`, or `onError`. There are also tests that verify proper permission handling by exercising both normal and exceptional coroutine cancellation.

Reply:
That's correct

Reply:
@RobWin Even if we catch and rethrow the exception in the flow body, that will only allow us to handle success / error cases. Using the onCompletion operator ensures that our event reporting is still invoked even in the case of cancellation.

Reply:
Your current implementation looks good to me